File: TallerProgra/Taller2023/ProgramacionDinamica/maxAdjDpRecursiva.cpp
#include <iostream>
using namespace std;

const int maxN = 1000000;
int n;
int arreglo[maxN];

int dp[maxN];
bool visited[maxN];
int mnss(int i){
    if( i > n-1)
        return 0;
    else{
        if( !visited[i] ){
            visited[i] = 1;
            //el maximo entre tomar y no tomar
            dp[i] = max(arreglo[i] + mnss(i+2), mnss(i+1));
        }
        return dp[i];
    }
}

int main(){
    cin>>n;
    for(int i = 0; i < n; ++i){
        cin>>arreglo[i];
    }
    cout<<mnss(0)<<"\n";
        
}

File: TallerProgra/Taller2023/ProgramacionDinamica/maxAdjDpRecursiva.py
import numpy as np
#maxN = 1000000

def mnss(index):
    if index >= n:
        return 0
    if not visited[index]:
        dp[index] = max(array[index] + mnss(index+2), mnss(index+1))
    return dp[index]

def getResp():
    resp = []
    i = 0
    while i < n:
        # lo uso 
        #ultimo del arreglo 
        if dp[i] == array[i] + (dp[i+2] if i+2 < n else 0):
            resp.append(i)
            i = i+2
        # no lo uso
        else:
            i = i+1
    return resp
        

n = int(input())
array = [int(num) for num in input().split(' ')]
visited = np.full([n], False)
dp = np.full([n], 0)
print(mnss(0))
print(getResp())

